eos-paygd/main.c

```diff
@@ -250,7 +250,7 @@ main (int   argc,
   g_option_context_add_main_entries (context, opts, GETTEXT_PACKAGE);
   if (!g_option_context_parse (context, &argc, &argv, &error))
     {
-      g_warning ("Failed to parse options: %s", error->message);
+      g_critical ("Failed to parse options: %s", error->message);
```

Comment:
> `g_critical()` will abort the program if run with `G_DEBUG=fatal-warnings`, which often happens in unit testing environments.
> 

I was aware of that behavior, but the same is true for `g_warning` as well, right? At least according to https://people.gnome.org/~ryanl/glib-docs/glib-running.html. And since we already use `g_warning` here (and even with this PR our testing suite still passes), I think this would not impact us.

> Its documentation says “Logging of a critical error is by definition an indication of a bug somewhere in the current program (or its libraries).”
> 

On https://people.gnome.org/~ryanl/glib-docs/glib-Message-Logging.html#g-critical it says differently: "It's more or less application-defined what constitutes a critical vs. a regular warning. You could call g_log_set_always_fatal() to make critical warnings exit the program, then use g_critical() for fatal errors, for example." Since we don't call `g_log_set_always_fatal` it seemed to me that the only difference is the log level severity used.

But I am not sure I'm looking at the correct up-to-date documentation (unfortunately it is the best result google gave me). Can you point me to the most accurate documentation online?

> For printing an error on failing to parse the input options, you probably want to use `g_printerr()` directly.

IIUC, `g_printerr` will print this message to stderr, which I agree makes sense for regular daemons, as it is good practice to tell the caller what is going on. I'm not sure if this is what we want for paygd, since it is only started by its unit file from the initrd (ignoring legacy installations, where it is started after pivot root) or during provisioning to query the security level (returned in a user-hostile binary format). ATM we don't print anything to stdout / stderr, all messaging goes through g_warning / g_message / g_debug (and this PR adds g_critical).

Reply:
I updated a couple of other fatal errors that I had missed from warning to critical as well -- if we ultimately decide against it we can revert all of them together.

Reply:
> On https://people.gnome.org/~ryanl/glib-docs/glib-Message-Logging.html#g-critical it says differently

Yeah, that documentation is waaaaaay old. The latest docs are at https://docs.gtk.org/glib/. They’ve been ported to a new documentation system and there are some glitches in places. If a glitch affects your enjoyment of the documentation, the previous version is at https://developer-old.gnome.org/glib/unstable/ (for GLib 2.68, which is now 12 months old).

Reply:
> IIUC, g_printerr will print this message to stderr, which I agree makes sense for regular daemons, as it is good practice to tell the caller what is going on. I'm not sure if this is what we want for paygd, since it is only started by its unit file from the initrd (ignoring legacy installations, where it is started after pivot root) or during provisioning to query the security level (returned in a user-hostile binary format). ATM we don't print anything to stdout / stderr, all messaging goes through g_warning / g_message / g_debug (and this PR adds g_critical).

That reasoning seems sound. I’d stick with `g_warning()` then, rather than using `g_critical()`.

It’s all a bit moot, given that this error message is printed if the command line options are invalid, which won’t ever happen when run from a systemd unit.

Reply:
Great, makes sense, I'll revert back to `g_warning` -- thanks for helping clarify all of this.